File: 剑指offer/63.股票的最大利润.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n == 0:
            return 0
        dp = [[0, 0] for _ in range(n)]

        for i in range(n):
            if i == 0:
                dp[i][0] = 0
                dp[i][1] = -prices[0]
            else:
                dp[i][0] = max(dp[i - 1][0], dp[i - 1][1] + prices[i])

                dp[i][1] = max(dp[i - 1][1], -prices[i])

        return dp[-1][0]

File: 剑指offer/63.股票的最大利润.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices)
    {
        int n = prices.size();

        if (n == 0) {
            return 0;
        }

        vector<vector<int>> dp(n, vector<int>(2));

        for (int i = 0; i < n; i++) {
            if (i == 0) {
                dp[0][1] = -prices[0];
                dp[0][0] = 0;
            } else {
                dp[i][0] = max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
                dp[i][1] = max(dp[i - 1][1], -prices[i]);
            }
        }

        return dp[n - 1][0];
    }
};